### DISCUSSION TITLE
q.dialog.onDismiss mixes with .onOk in an odd way

### DISCUSSION THREAD

**Discussion by chimmelb** [quasarframework/quasar#12636] [General - Components / Directives / etc]
I found that the chained `.onDismiss()` function was conflicting with `.onOk()` I am unsure if this is a bug or intended behavior, so am documenting it here.

[Here is a codepen to demonstrate](https://codepen.io/chimmelb/pen/gOXQzdq?editors=1011).

The setup is a dialog box with a prompt for user input. I'm using it to create a new object by name, calling my server API, and responding with the newly created object ID. To accomplish this, I wrap the `q.dialog()` in a Promise to resolve with the promises `resolve` function. 

**What should happen:** When I enter text and click ok, the API should run and resolve the promise with the new ID (in this Pen, the hard-coded ID is "abc123")

**What does happen:** The `onDismiss()` and `onOk()` fire when I click ok. This calls the API in onOK (just a timeout to simulate), but also runs onDismiss, which completes the promise. Not only that, onDismiss is called with the value of the prompt! (which should only be passed to onOK() - I would expect onDismiss to be called with `undefined` like when clicking outside.)

**Workaround:** Making the dialog persistent and removing the chained onDismiss(). Everything works as expected then, even wrapped in a Promise.

**Comment by yusufkandemir** [comment]
`onOk` + `onCancel` = `onDismiss`
So, since you are handling `onOk` and `onCancel` separately and resolving in both of them in all cases, you should avoid resolving inside `onDismiss`.

For the "`onDismiss` getting called with a parameter" part, yes it's what it's doing. According to the documented types, it shouldn't happen, it should be called without any parameters as you have said. So, please create an issue about it, so we can track it properly. But, in general, resolving explicitly is a better idea to avoid undocumented parameters getting passed around, or an unintended small change breaking your code, or a function introducing new parameters although you were not using/expecting them, etc. Stuff can happen in your own code, in external packages, or even the Browser APIs depending on the browser vendor or version.
